lib/instana/secrets.rb

```diff
@@ -0,0 +1,41 @@
+require 'uri'
+require 'cgi'
+
+module Instana
+  class Secrets    
+    def remove_from_query(str, secret_values = Instana.agent.secret_values)
+      return str unless secret_values
+      
+      url = URI(str)
+      params = CGI.parse(url.query)
+      
+      redacted = params.map do |k, v|
+        needs_redaction = secret_values['list']
+          .any? { |t| matcher(secret_values['matcher']).(t,k) }
+        [k, needs_redaction ? '<redacted>' : v]
+      end
+      
+      url.query = URI.encode_www_form(redacted)
+      CGI.unescape(url.to_s)
+    end
+    
+    private
+    
+    def matcher(name)
+      case name
+      when 'equals-ignore-case'
+        ->(expected, actual) { expected.casecmp(actual) == 0 }
+      when 'equals'
+        ->(expected, actual) { (expected <=> actual) == 0 }
+      when 'contains-ignore-case'
+        ->(expected, actual) { actual.downcase.include?(expected) }
+      when 'contains'
+        ->(expected, actual) { actual.include?(expected) }
+      when 'regex'
+        ->(expected, actual) { !Regexp.new(expected).match(actual).nil? }
+      else
+        fail '#{@matcher} is not supported.'
```

Comment:
Wouldn't this fail the whole request handling in case the secrets is not configured?

Reply:
The [agent ships](https://www.instana.com/docs/setup_and_manage/host_agent/configuration/#secrets) a default configuration for secrets. This will only fail the request if the secrets configuration passed to us contains a matcher which we do not handle.

If we feel that the request should go through even if the secrets config is off, the two options for this that come to mind are:

1.  Return `lambda { true }` and redact everything and log a warning.
2. Return `lambda { false }` and allow the url to go out send a notification to `com.instana.plugin.generic.event`.

Reply:
@basti1302 what do we do in other tracers when the secret configuration is invalid?

Reply:
Warn about the invalid config and move on with the default configuration as defined in the agent. (The Node.js package contains a copy of the default config for that purpose). 

https://github.com/instana/nodejs-sensor/blob/main/packages/core/src/util/normalizeConfig.js#L329-L353

Looks like Hunter has already found that and copied that behaviour.